### PR TITLE
Add form to admin rubric which allows to manually create deposit

### PR DESCRIPTION
**Pull request by ysv**

### CODE REVIEW

**Comment by ysv** [comment]
@yivo how we can process exceptions for example invalid currency code etc

**Comment by yivo** [comment]
@yssavchuk everything should be validates using activemodel validators.

**Comment by yivo** [comment]
I think this is not in the scope for this PR.

**Comment by ysv** [comment]
@yivo review please

File: app/models/currency.rb [hunk 1]
```diff
@@ -30,6 +30,10 @@ def coins
       @coins ||= Currency.where(coin: true)
     end
 
+    def fiats
```
Comment (yivo) [inline app/models/currency.rb:33]: @yssavchuk pls refactor this to scope.

File: app/views/admin/deposits/banks/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        <div class="panel-body">
+            <dl class="dl-horizontal dl-member">
+              <dt><%= 'Bank' %></dt>
+              <dd><%= f.text_field :fund_extra, class: 'col-xs-24' %></dd>
+              <dt><%= 'Account' %></dt>
+              <dd><%= f.text_field :fund_uid, class: 'col-xs-24' %></dd>
+              <hr class="split">
+              <dt><%= 'Amount' %></dt>
+              <dd><%= f.number_field :amount, class: 'col-xs-24' %></dd>
+              <!--TODO: Currency.fiats.map(&:code)-->
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk Not ready?

File: app/views/admin/deposits/banks/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+              <dt><%= 'Currency' %></dt>
+              <dd><%= f.collection_radio_buttons :currency, [['USD'],['EUR'],['UAH']], :first, :last %></dd>
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk TODO

File: app/views/admin/deposits/banks/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+          <span><%= 'User info' %></span>
+        </div>
+        <div class="panel-body">
+            <dl class="dl-horizontal dl-member">
+              <%= f.fields_for :member, @bank.member do |p| %>
+                <dt><%= 'SN' %></dt>
+                <dd><%= p.text_field :sn%></dd>
+              <% end %>
+              <hr class="split">
+              <%= f.submit 'Accept', class: 'btn btn-primary pull-right'%>
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk Can you add space before `%>`?

File: app/views/admin/deposits/banks/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+              <dd><%= f.collection_radio_buttons :currency, [['USD'],['EUR'],['UAH']], :first, :last %></dd>
+            </dl>
+        </div>
+      </div>
+    </div>
+
+    <div class="col-xs-10">
+      <div class="panel panel-info">
+        <div class="panel-heading">
+          <span><%= 'User info' %></span>
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk Why ERB is needed here?

File: app/views/admin/deposits/banks/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        </div>
+        <div class="panel-body">
+            <dl class="dl-horizontal dl-member">
+              <%= f.fields_for :member, @bank.member do |p| %>
+                <dt><%= 'SN' %></dt>
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk Why ERB is needed here?

File: app/views/admin/deposits/banks/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+            <dl class="dl-horizontal dl-member">
+              <dt><%= 'Bank' %></dt>
+              <dd><%= f.text_field :fund_extra, class: 'col-xs-24' %></dd>
+              <dt><%= 'Account' %></dt>
+              <dd><%= f.text_field :fund_uid, class: 'col-xs-24' %></dd>
+              <hr class="split">
+              <dt><%= 'Amount' %></dt>
+              <dd><%= f.number_field :amount, class: 'col-xs-24' %></dd>
+              <!--TODO: Currency.fiats.map(&:code)-->
+              <dt><%= 'Currency' %></dt>
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk Why ERB is needed here?

File: app/views/admin/deposits/banks/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+          <span><%= 'Deposit info' %></span>
+        </div>
+        <div class="panel-body">
+            <dl class="dl-horizontal dl-member">
+              <dt><%= 'Bank' %></dt>
+              <dd><%= f.text_field :fund_extra, class: 'col-xs-24' %></dd>
+              <dt><%= 'Account' %></dt>
+              <dd><%= f.text_field :fund_uid, class: 'col-xs-24' %></dd>
+              <hr class="split">
+              <dt><%= 'Amount' %></dt>
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk Why ERB is needed here?

File: app/views/admin/deposits/banks/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    <div class="col-xs-10">
+      <div class="panel panel-primary">
+        <div class="panel-heading">
+          <span><%= 'Deposit info' %></span>
+        </div>
+        <div class="panel-body">
+            <dl class="dl-horizontal dl-member">
+              <dt><%= 'Bank' %></dt>
+              <dd><%= f.text_field :fund_extra, class: 'col-xs-24' %></dd>
+              <dt><%= 'Account' %></dt>
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk Why ERB is needed here?

File: app/views/admin/deposits/banks/new.html.erb [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+  <%= form_for [:admin, @bank] do |f| %>
+    <div class="col-xs-10">
+      <div class="panel panel-primary">
+        <div class="panel-heading">
+          <span><%= 'Deposit info' %></span>
+        </div>
+        <div class="panel-body">
+            <dl class="dl-horizontal dl-member">
+              <dt><%= 'Bank' %></dt>
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk Why ERB is needed here?

File: app/views/admin/deposits/banks/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,45 @@
+<div class="row">
+
+  <%= form_for [:admin, @bank] do |f| %>
+    <div class="col-xs-10">
+      <div class="panel panel-primary">
+        <div class="panel-heading">
+          <span><%= 'Deposit info' %></span>
```
Comment (yivo) [inline app/views/admin/deposits/banks/new.html.erb]: @yssavchuk Why ERB is needed here?

File: app/models/concerns/fund_sourceable.rb [hunk 12]
```diff
@@ -4,11 +4,11 @@ module FundSourceable
   included do
     attr_accessor :fund_source_id
     before_validation :set_fund_source_attributes, on: :create
-    validates :fund_source_id, presence: true, on: :create
+    validates :fund_extra, :fund_uid, presence: true, on: :create
   end
 
   def set_fund_source_attributes
-    if fund_source = FundSource.find_by(id: fund_source_id)
+    if fund_extra.nil? && fund_uid.nil? && fund_source = FundSource.find_by(id: fund_source_id)
```
Comment (yivo) [inline app/models/concerns/fund_sourceable.rb:11]: @yssavchuk Can you explain what fund_extra, fund_uid are?

File: app/models/concerns/fund_sourceable.rb [hunk 13]
```diff
@@ -4,11 +4,11 @@ module FundSourceable
   included do
     attr_accessor :fund_source_id
     before_validation :set_fund_source_attributes, on: :create
-    validates :fund_source_id, presence: true, on: :create
+    validates :fund_extra, :fund_uid, presence: true, on: :create
```
Comment (yivo) [inline app/models/concerns/fund_sourceable.rb:7]: @yssavchuk Can you explain the changes?

File: app/controllers/admin/deposits/banks_controller.rb [hunk 14]
```diff
@@ -20,6 +22,11 @@ def show
         flash.now[:notice] = t('.notice') if @bank.aasm_state.accepted?
       end
 
+      def create
+        ::Deposits::Bank.create!(deposit_params)
```
Comment (yivo) [inline app/controllers/admin/deposits/banks_controller.rb]: @yssavchuk Can you display errors? Something like: `if deposit.save then redirect else display errors`.

Comment (yivo) [reply app/controllers/admin/deposits/banks_controller.rb]: Typical for Rails.

File: app/controllers/admin/deposits/banks_controller.rb [hunk 15]
```diff
@@ -35,6 +42,25 @@ def update
...
+
+      def deposit_params
+        params.require(:deposits_bank).permit(:currency, :amount, :fund_uid, :fund_extra, :member_id, :account_id)
+      end
+
+      def member_params
+        params.require(:deposits_bank).require(:member)
+      end
+
+      def fetch_member!
```
Comment (yivo) [inline app/controllers/admin/deposits/banks_controller.rb]: @yssavchuk I don't like this behavior. Let's create `def sn=(new_sn)` at bank deposit model. Insite the method we can look up member by SN and `self.member = member`. There should be validation on deposit for presence of member so if no member is available deposit will notify about error and could not be saved. Also try to omit modifying params.

File: app/controllers/admin/deposits/banks_controller.rb [hunk 16]
```diff
@@ -20,6 +22,11 @@ def show
         flash.now[:notice] = t('.notice') if @bank.aasm_state.accepted?
       end
 
+      def create
+        ::Deposits::Bank.create!(deposit_params)
+        redirect_to admin_deposits_banks_path
```
Comment (yivo) [inline app/controllers/admin/deposits/banks_controller.rb]: @yssavchuk Can we use redirect_to :back?

**Review by yivo — CHANGES_REQUESTED** [review]

**Comment by yivo** [comment]
@yssavchuk done

File: app/models/concerns/fund_sourceable.rb (hunk 13 above)
Comment (ysv) [reply app/models/concerns/fund_sourceable.rb:7]: @yivo actually we don't need fund_source (which represents bank account) to create bank deposit all we need  is fund_extra(bank account) & fund_uid(user account in bank)

File: app/models/concerns/fund_sourceable.rb (hunk 12 above)
Comment (ysv) [reply app/models/concerns/fund_sourceable.rb:11]: fund_extra - sender bank name
fund_uid - user account in bank

File: app/models/currency.rb (hunk 1 above)
Comment (ysv) [reply app/models/currency.rb:33]: @yivo we cant because ActiveYmlBase doesn't provide such method

**Review by yivo — APPROVED** [review]

**Pull request merged**